### ISSUE TITLE
Semantics of `convert`

### ISSUE THREAD

**Issue by TotalVerb** [JuliaLang/julia#18270]
From https://github.com/JuliaLang/julia/issues/18264#issuecomment-242953334:

I really think there should be some well-defined semantics (axioms, if you will) for `convert`:
- It should be bidirectional; `oftype(y, oftype(x, y))` should either be `y` or an error (with minor exceptions of `0.0` vs `-0.0`, etc.). There are a couple of cases where this is violated right now, and I think those should be considered bugs.
- Converted value should in some sense be the "same" as pre-converted value. In most cases `x == convert(T, x)` holds; but this is not true in this case, and also in some other cases like `convert(String, ::Vector{Char})`. Rather I guess this sense of "same"ness is a little more loose.

cc @nalimilan

**Comment by toivoh** [comment]
How about converting a `Float64` to a `Float32` and then back again? I don't think that should be an error even when the roundtrip is not exact.

**Comment by TotalVerb** [comment]
Good point — but that does raise the question of what kind of rounding is allowable. For instance, we allow `Rational{Int32}` to be converted to `Float64`, losing precision without `InexactError`. But when converting from `Float64` to `Rational{Int32}`, the current behaviour is to raise an `InexactError`.

**Comment by StefanKarpinski** [comment]
I agree that this should be spelled out, but I think it's primarily a documentation issue. Insisting on perfect round tripping would be too annoying in practice – it is very common to convert values to floating-point and lose some precision. The general concept is that some types are exact (integers, rationals) while other types are approximate (floating points). If you convert to an approximate type, you may lose some precision and an error isn't raised. If you convert to an exact type, precision is not lost and an error is raised if precision cannot be preserved.

**Comment by TotalVerb** [comment]
Another question (not to be determined in `Base`, necessarily, but `Base` does have `Dates` which has this issue) is whether `convert` should be able to remove or add units. For instance, currently

```julia
julia> convert(Int, Dates.Second(10))
10
```

This is convenient syntax, but at the same time, implicitly removing the unit in some cases seems to me as bad as unwrapping nullables, which concerns were raised about in https://github.com/JuliaLang/julia/issues/18264#issuecomment-242953334.

**Comment by timholy** [comment]
Unit-stripping and addition should not be allowed, because then 2-step conversions can do things that you'd probably never permit with a 1-step conversion. https://github.com/JuliaLang/julia/pull/17611#issuecomment-235048985

**Comment by TotalVerb** [comment]
Here's another situation I came about recently while working on a traits-based library.

Is `convert` suitable to convert to a trait? That is, should something like

```julia
convert(::Type{Base.LinearFast}, x) = ...
```

be a valid `convert` definition, or should this be made into a separate method? My leaning is toward separate method, since this would break the (usual) convention that `SomeType(x)::SomeType`.

**Comment by JeffBezanson** [comment]
Yes, it should be a separate method, since you're not returning something of type `LinearFast`. `LinearFast(x)` would be better than `convert` for this; less code makes assumptions about the result of constructors than about convert.

**Comment by TotalVerb** [comment]
Should this be closed by the latest decisions on `convert` vs. type constructors?

**Comment by JeffBezanson** [comment]
Yes, this will be closed by #23273, which includes a new description of `convert` in the manual.

**Comment by JeffBezanson** [comment]
Closed by #23273.
